app/src/org/commcare/android/util/SigningUtil.java

```diff
@@ -0,0 +1,181 @@
+package org.commcare.android.util;
+
+import org.commcare.android.logic.GlobalConstants;
+import org.spongycastle.jce.provider.BouncyCastleProvider;
+
+import java.io.UnsupportedEncodingException;
+import java.security.InvalidKeyException;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.security.spec.X509EncodedKeySpec;
+
+/**
+ * A set of helper methods for verifying whether a message was genuinely sent from HQ. Currently we
+ * expcect the SMS in the format [COMMCARE APP - DO NOT DELETE] {BASE64-Encoded-Message} EG
+ *
+ * [COMMCARE APP - DO NOT DELETE] Y2NhcHA6IGh0dHA6Ly9iaXQubHkvMU5JMUl6MyBzaWduYXR1cmU6IEvECygFUhiUH
+ * 3TRjC0lClQrpLR7lG//IpDYpRH7ComtZRjTirteXmPyM9fRgbPZ9K6jG9zEms9WQj55Uo7jTujKNYThjU8rJJmWLouJBr/Yn
+ * WobEupwzn6DP2FavPF1YLPbp0ZctOfymW3m4j3VZ0lR2dMOjmInMSBiInqICKid
+ *
+ * WHERE BASE64-Encoded-Message is decoded into:
+ *
+ * ccapp: <profile link> signature: <binary signature>
+ *
+ * And we can then verify that the profile link was in fact signed (using SHA256withRSA) by
+ * the CommCareHQ private key
+ *
+ * Created by wpride1 on 9/11/15.
+ */
+public class SigningUtil {
+    /**
+     *
+     * @param text the parsed out text message in the expected link/signature format
+     * @return the download link if the message was valid and verified, null otherwise
+     * @throws SignatureException if we discovered a valid-looking message but could not verifyMessageSignatureHelper it
+     */
+    public static String parseAndVerifySMS(String text) throws SignatureException {
+        // parse out the app link and signature. We assume there is a space after ccapp: and
+        // signature: and that the end of the signature is the end of the text content
+        try {
+            byte[] signatureBytes = getSignatureBytes(text);
+            String decodedMessage = parseAndDecodeSMS(text);
```

Comment:
The `parseAndDecodeSMS` method still calls `decodeEncodedSMS` which calls `String()` on the whole byte array which will cause the issue I mentioned earlier.  It looks like you just have to make `decodeEncodedSMS` call `String()` on only the the `ccapp: ... signature:` part of the message (i.e., excluding the binary signature) and then this should be good.